### ISSUE TITLE
`possibly-undefined` false positive when assigning variable for every case in match statement for all possible values

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#16992]
```py
# mypy: enable-error-code=possibly-undefined

from typing import Literal

value: Literal[1, 2]

match value:
    case 1:
        result = 1
    case 2:
        result = 2

result # error: possibly-undefined
```
[playground](https://mypy-play.net/?mypy=master&python=3.12&flags=strict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Clocal-partial-types%2Cstrict-equality%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=7ada949d8f753dab131a7a9a6076a9f4)

similar issue: #15958 (seems it was only fixed when using `assert_never` at the end)

**Comment by Andrej730** [comment]
Interesting, that adding `case _` *kind of* resolves the issue though mypy does detect that `case _` code is unreachable.

```python
# mypy: enable-error-code=possibly-undefined

from typing import Literal

value: Literal[1, 2]

match value:
    case 1:
        result = 1
    case 2:
        result = 2
    case _:
        # error: Statement is unreachable  [unreachable]
        result = 3

result
```


I've also met the similar issue with `if isinstance` checks. In the last case adding redudant `else` block makes the case false negative.
```python
# mypy: enable-error-code="possibly-undefined"

class A: ...
class B: ...

# False positive: Name "new_name" may be undefined  [possibly-undefined]
def test_1(obj: A | B) -> None:
    if isinstance(obj, A):
        new_name = "test"
    elif isinstance(obj, B):
        new_name = "test"
    a = new_name

# + adding "else: reveal_type(obj)" somehow resolves it
def test_1b(obj: A | B) -> None:
    if isinstance(obj, A):
        new_name = "test"
    elif isinstance(obj, B):
        new_name = "test"
    else:
        reveal_type(obj)
    a = new_name

# False positive: Name "new_name" may be undefined  [possibly-undefined]
def get_obj() -> A | B: ...
def test_2() -> None:
    obj = get_obj()
    if isinstance(obj, A):
        new_name = "test"
    elif isinstance(obj, B):
        new_name = "test"
    a = new_name
# but adding "else: reveal_type(obj)" somehow resolves it
def test_2b() -> None:
    obj = get_obj()
    if isinstance(obj, A):
        new_name = "test"
    elif isinstance(obj, B):
        new_name = "test"
    else:
        reveal_type(obj)
    a = new_name

# Name "new_name" may be undefined  [possibly-undefined]
# and it's correct
def test_3(obj: str | float) -> None:
    if isinstance(obj, str):
        new_name = "test"
    elif isinstance(obj, float):
        new_name = "test"
    a = new_name

# but adding "else: reveal_type(obj)" makes it false negative
def test_3b(obj: str | float) -> None:
    if isinstance(obj, str):
        new_name = "test"
    elif isinstance(obj, float):
        new_name = "test"
    else:
        reveal_type(obj)
    a = new_name
```

**Comment by tamird** [comment]
Another simple case that demonstrates the inconsistency with unreachable code:

```python
def takes_list_problem[T](stuff: list[T]) -> None:
    match stuff:
        case [one]:
            pass
        case [*_]:
            raise AssertionError("nope")
            
    print(one) # error: Name "one" may be undefined  [possibly-undefined]
    
def takes_list_no_proble[T](stuff: list[T]) -> None:
    def one() -> T:
        match stuff:
            case [one]:
                return one
            case [*_]:
                raise AssertionError("nope")
    
    print(one())
```

[playground](https://mypy-play.net/?mypy=1.20.2&python=3.12&enable-error-code=possibly-undefined%2Cexhaustive-match&flags=warn-unreachable&gist=59be7937e13fc2b945bd251bcfed21ec)
